File: file_log.py
import os
import datetime
from colorama import Fore, Style, init

# Initialize colorama
init(autoreset=True)

# File icons by extension
FILE_ICONS = {
    '.py': '🐍',
    '.csv': '📊',
    '.md': '📝',
    '.txt': '📝',
    '.jpg': '🖼️',
    '.jpeg': '🖼️',
    '.png': '🖼️',
    '.gif': '🖼️',
    '.pdf': '📄',
    '.zip': '📦',
    '.tar': '📦',
    '.gz': '📦',
}

def format_size(bytes_size):
    """Format bytes into KB, MB, or GB."""
    for unit in ['B', 'KB', 'MB', 'GB', 'TB']:
        if bytes_size < 1024:
            return f"{bytes_size:.2f} {unit}"
        bytes_size /= 1024
    return f"{bytes_size:.2f} PB"

def format_date(timestamp):
    """Format timestamp into human-readable date."""
    return datetime.datetime.fromtimestamp(timestamp).strftime('%Y-%m-%d %H:%M:%S')

def get_file_icon(filename):
    """Return icon based on file extension."""
    ext = os.path.splitext(filename)[1].lower()
    return FILE_ICONS.get(ext, '📄')

def scan_directory_recursive(directory='.'):
    """Recursively list all files, skipping dot-directories, in a colorful tree view with icons."""
    print(f"\n{Fore.YELLOW}📂 Scanning directory: {os.path.abspath(directory)}")
    print("=" * 90)

    for root, dirs, files in os.walk(directory):
        # Skip dot-directories
        dirs[:] = [d for d in dirs if not d.startswith('.')]
        level = root.replace(directory, '').count(os.sep)
        indent = '│   ' * level + '├── ' if level > 0 else ''
        print(f"{Fore.BLUE}{indent}📁 {os.path.basename(root)}/")
        sub_indent = '│   ' * (level + 1)
        for f in files:
            file_path = os.path.join(root, f)
            try:
                size = format_size(os.path.getsize(file_path))
                modified = format_date(os.path.getmtime(file_path))
                icon = get_file_icon(f)
                print(f"{sub_indent}{Fore.GREEN}{icon} {f} "
                      f"{Fore.YELLOW}- {size} "
                      f"{Fore.MAGENTA}- {modified}{Style.RESET_ALL}")
            except Exception as e:
                print(f"{sub_indent}{Fore.RED}⚠️ {f} (Error: {e})")

    print(f"\n{Fore.GREEN}🌲 Scan complete!\n")

if __name__ == '__main__':
    directory_path = input("Enter directory path (press Enter for current directory): ").strip()
    if not directory_path:
        directory_path = '.'
    scan_directory_recursive(directory_path)

File: file_log.cpp
// file_log.cpp - C++ version of file_log.py
#include <iostream>
#include <filesystem>
#include <iomanip>
#include <chrono>
#include <map>
#include <string>

namespace fs = std::filesystem;

// File icons by extension
std::map<std::string, std::string> FILE_ICONS = {
    {".py", "🐍"}, {".csv", "📊"}, {".md", "📝"}, {".txt", "📝"},
    {".jpg", "🖼️"}, {".jpeg", "🖼️"}, {".png", "🖼️"}, {".gif", "🖼️"},
    {".pdf", "📄"}, {".zip", "📦"}, {".tar", "📦"}, {".gz", "📦"}
};

std::string format_size(uintmax_t bytes) {
    const char* units[] = {"B", "KB", "MB", "GB", "TB", "PB"};
    double size = bytes;
    int unit_index = 0;
    while (size >= 1024 && unit_index < 5) {
        size /= 1024;
        ++unit_index;
    }
    std::ostringstream oss;
    oss << std::fixed << std::setprecision(2) << size << " " << units[unit_index];
    return oss.str();
}

std::string format_date(const std::filesystem::file_time_type& ftime) {
    auto sctp = std::chrono::time_point_cast<std::chrono::system_clock::duration>(ftime - fs::file_time_type::clock::now()
                            + std::chrono::system_clock::now());
    std::time_t cftime = std::chrono::system_clock::to_time_t(sctp);
    std::ostringstream oss;
    oss << std::put_time(std::localtime(&cftime), "%Y-%m-%d %H:%M:%S");
    return oss.str();
}

std::string get_file_icon(const std::string& filename) {
    std::string ext = fs::path(filename).extension().string();
    for (auto& c : ext) c = tolower(c);
    return FILE_ICONS.count(ext) ? FILE_ICONS[ext] : "📄";
}

void scan_directory_recursive(const std::string& directory) {
    std::cout << "\n📂 Scanning directory: " << fs::absolute(directory) << "\n";
    std::cout << std::string(90, '=') << "\n";

    for (const auto& entry : fs::recursive_directory_iterator(directory)) {
        std::string indent(entry.depth() * 4, ' ');
        if (entry.is_directory()) {
            std::cout << indent << "📁 " << entry.path().filename().string() << "/\n";
        } else {
            try {
                std::string icon = get_file_icon(entry.path().string());
                std::string size = format_size(fs::file_size(entry.path()));
                std::string modified = format_date(fs::last_write_time(entry.path()));
                std::cout << indent << icon << " " << entry.path().filename().string()
                          << " - " << size << " - " << modified << "\n";
            } catch (const std::exception& e) {
                std::cerr << indent << "⚠️ " << entry.path().filename().string()
                          << " (Error: " << e.what() << ")\n";
            }
        }
    }
    std::cout << "\n🌲 Scan complete!\n";
}

int main() {
    std::string path;
    std::cout << "Enter directory path (press Enter for current directory): ";
    std::getline(std::cin, path);
    if (path.empty()) path = ".";
    scan_directory_recursive(path);
    return 0;
}
